File: server/internal/domain/weapons.go
package domain

// ============================================================
// v14 Phase 2: Megabonk Weapon System — Data & Types
// ============================================================

// WeaponType identifies one of the 10 weapon types.
type WeaponType string

const (
	WeaponBonkMallet    WeaponType = "bonk_mallet"
	WeaponChainBolt     WeaponType = "chain_bolt"
	WeaponFlameRing     WeaponType = "flame_ring"
	WeaponFrostShards   WeaponType = "frost_shards"
	WeaponShadowStrike  WeaponType = "shadow_strike"
	WeaponThunderClap   WeaponType = "thunder_clap"
	WeaponVenomCloud    WeaponType = "venom_cloud"
	WeaponCrystalShield WeaponType = "crystal_shield"
	WeaponGravityBomb   WeaponType = "gravity_bomb"
	WeaponSoulDrain     WeaponType = "soul_drain"
)

// AllWeaponTypes lists all weapon types for iteration.
var AllWeaponTypes = []WeaponType{
	WeaponBonkMallet, WeaponChainBolt, WeaponFlameRing, WeaponFrostShards,
	WeaponShadowStrike, WeaponThunderClap, WeaponVenomCloud, WeaponCrystalShield,
	WeaponGravityBomb, WeaponSoulDrain,
}

// WeaponPattern describes the firing pattern type.
type WeaponPattern string

const (
	PatternFanSwing     WeaponPattern = "fan_swing"      // 전방 부채꼴
	PatternChainTarget  WeaponPattern = "chain_target"    // 체인 연쇄
	PatternCircularRing WeaponPattern = "circular_ring"   // 360° 원형
	PatternMultiShot    WeaponPattern = "multi_shot"      // 전방 다발
	PatternTeleportBack WeaponPattern = "teleport_back"   // 배후 텔레포트
	PatternTargetedAOE  WeaponPattern = "targeted_aoe"    // 타겟 AOE
	PatternDeployable   WeaponPattern = "deployable"      // 설치형
	PatternOrbital      WeaponPattern = "orbital"         // 궤도 회전
	PatternDeployExplode WeaponPattern = "deploy_explode" // 설치 폭발
	PatternBeam         WeaponPattern = "beam"            // 빔
)

// SpecialEffect describes the secondary effect of a weapon.
type SpecialEffect string

const (
	EffectNone      SpecialEffect = "none"
	EffectKnockback SpecialEffect = "knockback"
	EffectChainDecay SpecialEffect = "chain_decay"
	EffectBurnDOT   SpecialEffect = "burn_dot"
	EffectSlow      SpecialEffect = "slow"
	EffectBackstab  SpecialEffect = "backstab"
	EffectStun      SpecialEffect = "stun"
	EffectPoisonDOT SpecialEffect = "poison_dot"
	EffectReflect   SpecialEffect = "reflect"
	EffectPull      SpecialEffect = "pull"
	EffectLifesteal SpecialEffect = "lifesteal"
)

// WeaponData defines the static properties of a weapon at base (Lv1).
type WeaponData struct {
	Type          WeaponType    `json:"type"`
	Name          string        `json:"name"`
	Description   string        `json:"description"`
	BaseDPS       float64       `json:"baseDPS"`       // damage per hit
	Range         float64       `json:"range"`         // effective range in px
	CooldownSec   float64       `json:"cooldownSec"`   // seconds between attacks
	Pattern       WeaponPattern `json:"pattern"`
	SpecialEffect SpecialEffect `json:"specialEffect"`
	// Pattern-specific parameters
	FanAngleDeg    float64 `json:"fanAngleDeg,omitempty"`    // BonkMallet: 120°
	ChainCount     int     `json:"chainCount,omitempty"`     // ChainBolt: 3
	ChainDecay     float64 `json:"chainDecay,omitempty"`     // ChainBolt: -20% per chain
	ProjectileCount int    `json:"projectileCount,omitempty"` // FrostShards: 5
	KnockbackPx    float64 `json:"knockbackPx,omitempty"`    // BonkMallet: 30px
	DOTDurationSec float64 `json:"dotDurationSec,omitempty"` // FlameRing/VenomCloud: seconds
	DOTDPSPerSec   float64 `json:"dotDPSPerSec,omitempty"`   // DOT damage per second
	SlowPercent    float64 `json:"slowPercent,omitempty"`     // FrostShards: 0.40
	SlowDurationSec float64 `json:"slowDurationSec,omitempty"` // FrostShards: 1s
	StunDurationSec float64 `json:"stunDurationSec,omitempty"` // ThunderClap: 0.5s
	DeployDurationSec float64 `json:"deployDurationSec,omitempty"` // VenomCloud: 5s
	OrbitalCount   int     `json:"orbitalCount,omitempty"`   // CrystalShield: 3
	PullDurationSec float64 `json:"pullDurationSec,omitempty"` // GravityBomb: 2s
	LifestealPct   float64 `json:"lifestealPct,omitempty"`   // SoulDrain: 0.30
	BackstabMult   float64 `json:"backstabMult,omitempty"`   // ShadowStrike: 2.0
}

// WeaponEvolutionData defines the multipliers applied at each evolution level.
type WeaponEvolutionData struct {
	Level       int     `json:"level"`       // 1~5
	DPSMult     float64 `json:"dpsMult"`     // damage multiplier
	RangeMult   float64 `json:"rangeMult"`   // range multiplier
	CooldownMult float64 `json:"cooldownMult"` // cooldown multiplier (< 1 = faster)
	UltimateName string  `json:"ultimateName,omitempty"` // Lv5 only
	UltimateDesc string  `json:"ultimateDesc,omitempty"` // Lv5 only
}

// WeaponSlot holds a weapon instance in an agent's inventory.
type WeaponSlot struct {
	Type          WeaponType `json:"type"`
	Level         int        `json:"level"` // 1~5
	CooldownTicks int       `json:"cooldownTicks"` // remaining cooldown ticks
}

// MaxWeaponSlots is the maximum number of weapons an agent can hold.
const MaxWeaponSlots = 5

// MaxWeaponLevel is the maximum evolution level.
const MaxWeaponLevel = 5

// DamageSourceWeapon extends DamageSource for weapon-based deaths.
const DamageSourceWeapon DamageSource = "weapon"

// ============================================================
// Weapon Definitions — 10 Weapons
// ============================================================

// AllWeapons defines the base stats for all 10 weapons.
var AllWeapons = []WeaponData{
	{
		Type: WeaponBonkMallet, Name: "Bonk Mallet",
		Description:   "Smashes enemies in a 120° frontal cone with knockback",
		BaseDPS: 25, Range: 80, CooldownSec: 1.5,
		Pattern: PatternFanSwing, SpecialEffect: EffectKnockback,
		FanAngleDeg: 120, KnockbackPx: 30,
	},
	{
		Type: WeaponChainBolt, Name: "Chain Bolt",
		Description:   "Lightning bolt chains through 3 enemies with decreasing damage",
		BaseDPS: 15, Range: 150, CooldownSec: 1.2,
		Pattern: PatternChainTarget, SpecialEffect: EffectChainDecay,
		ChainCount: 3, ChainDecay: 0.20,
	},
	{
		Type: WeaponFlameRing, Name: "Flame Ring",
		Description:   "Expanding 360° ring of fire that burns enemies",
		BaseDPS: 20, Range: 120, CooldownSec: 3.0,
		Pattern: PatternCircularRing, SpecialEffect: EffectBurnDOT,
		DOTDurationSec: 2.0, DOTDPSPerSec: 10,
	},
	{
		Type: WeaponFrostShards, Name: "Frost Shards",
		Description:   "Fires 5 ice shards in a 45° cone, slowing enemies",
		BaseDPS: 12, Range: 180, CooldownSec: 1.0,
		Pattern: PatternMultiShot, SpecialEffect: EffectSlow,
		FanAngleDeg: 45, ProjectileCount: 5,
		SlowPercent: 0.40, SlowDurationSec: 1.0,
	},
	{
		Type: WeaponShadowStrike, Name: "Shadow Strike",
		Description:   "Teleports behind the nearest enemy for a backstab",
		BaseDPS: 40, Range: 200, CooldownSec: 4.0,
		Pattern: PatternTeleportBack, SpecialEffect: EffectBackstab,
		BackstabMult: 2.0,
	},
	{
		Type: WeaponThunderClap, Name: "Thunder Clap",
		Description:   "AOE lightning on the highest-HP enemy, stunning targets",
		BaseDPS: 30, Range: 250, CooldownSec: 3.5,
		Pattern: PatternTargetedAOE, SpecialEffect: EffectStun,
		StunDurationSec: 0.5,
	},
	{
		Type: WeaponVenomCloud, Name: "Venom Cloud",
		Description:   "Deploys a poison cloud at current position",
		BaseDPS: 8, Range: 100, CooldownSec: 5.0,
		Pattern: PatternDeployable, SpecialEffect: EffectPoisonDOT,
		DeployDurationSec: 5.0, DOTDurationSec: 5.0, DOTDPSPerSec: 8,
	},
	{
		Type: WeaponCrystalShield, Name: "Crystal Shield",
		Description:   "3 orbiting crystals that damage on contact and reflect projectiles",
		BaseDPS: 10, Range: 50, CooldownSec: 0,
		Pattern: PatternOrbital, SpecialEffect: EffectReflect,
		OrbitalCount: 3,
	},
	{
		Type: WeaponGravityBomb, Name: "Gravity Bomb",
		Description:   "Deploys a black hole that pulls enemies in, then explodes",
		BaseDPS: 35, Range: 200, CooldownSec: 6.0,
		Pattern: PatternDeployExplode, SpecialEffect: EffectPull,
		PullDurationSec: 2.0,
	},
	{
		Type: WeaponSoulDrain, Name: "Soul Drain",
		Description:   "Continuous beam to nearest enemy, stealing HP",
		BaseDPS: 18, Range: 120, CooldownSec: 0,
		Pattern: PatternBeam, SpecialEffect: EffectLifesteal,
		LifestealPct: 0.30,
	},
}

// GetWeaponData returns the WeaponData for a given WeaponType, or nil if not found.
func GetWeaponData(wt WeaponType) *WeaponData {
	for i := range AllWeapons {
		if AllWeapons[i].Type == wt {
			return &AllWeapons[i]
		}
	}
	return nil
}

// ============================================================
// Weapon Evolution — Lv1~5
// ============================================================

// WeaponEvolutionTable maps each level to its multipliers.
// Lv1 = base, Lv2 = +30% DMG, Lv3 = +25% range, Lv4 = -20% CD, Lv5 = ultimate.
var WeaponEvolutionTable = []WeaponEvolutionData{
	{Level: 1, DPSMult: 1.0, RangeMult: 1.0, CooldownMult: 1.0},
	{Level: 2, DPSMult: 1.3, RangeMult: 1.0, CooldownMult: 1.0},
	{Level: 3, DPSMult: 1.3, RangeMult: 1.25, CooldownMult: 1.0},
	{Level: 4, DPSMult: 1.3, RangeMult: 1.25, CooldownMult: 0.80},
	{Level: 5, DPSMult: 1.6, RangeMult: 1.5, CooldownMult: 0.70},
}

// GetEvolutionData returns the evolution multipliers for a given level (1-5).
func GetEvolutionData(level int) *WeaponEvolutionData {
	if level < 1 || level > MaxWeaponLevel {
		return nil
	}
	return &WeaponEvolutionTable[level-1]
}

// UltimateWeaponNames maps each weapon to its Lv5 ultimate transformation name.
var UltimateWeaponNames = map[WeaponType]string{
	WeaponBonkMallet:    "Earthquake",
	WeaponChainBolt:     "Storm Network",
	WeaponFlameRing:     "Inferno",
	WeaponFrostShards:   "Blizzard",
	WeaponShadowStrike:  "Phantom Dance",
	WeaponThunderClap:   "Judgment",
	WeaponVenomCloud:    "Plague",
	WeaponCrystalShield: "Diamond Fortress",
	WeaponGravityBomb:   "Singularity",
	WeaponSoulDrain:     "Life Siphon",
}

// UltimateWeaponDescs maps each weapon to its Lv5 ultimate description.
var UltimateWeaponDescs = map[WeaponType]string{
	WeaponBonkMallet:    "360° shockwave + crack zone",
	WeaponChainBolt:     "5-chain + residual lightning field",
	WeaponFlameRing:     "Double ring + piercing",
	WeaponFrostShards:   "360° ice storm",
	WeaponShadowStrike:  "Triple teleport + clone",
	WeaponThunderClap:   "Map-wide lightning + 5s stun (once per epoch)",
	WeaponVenomCloud:    "Moving cloud + infection spread",
	WeaponCrystalShield: "5 orbitals + 200% reflect",
	WeaponGravityBomb:   "5s black hole + 2x radius",
	WeaponSoulDrain:     "Frontal cone + ally heal",
}

// ============================================================
// Status Effects for Weapon System
// ============================================================

// StatusEffectType identifies a status effect applied by weapons.
type StatusEffectType string

const (
	StatusBurn     StatusEffectType = "burn"
	StatusPoison   StatusEffectType = "poison"
	StatusSlow     StatusEffectType = "slow"
	StatusStun     StatusEffectType = "stun"
	StatusPull     StatusEffectType = "pull"
	StatusKnockback StatusEffectType = "knockback_effect"
)

// StatusEffect represents an active status effect on an agent.
type StatusEffect struct {
	Type         StatusEffectType `json:"type"`
	SourceID     string           `json:"sourceId"`     // who applied it
	TicksLeft    int              `json:"ticksLeft"`     // remaining duration in ticks
	DamagePerTick float64         `json:"damagePerTick"` // for DOTs
	SlowFraction float64          `json:"slowFraction"`  // for slow (0.0-1.0)
	KnockbackX   float64          `json:"knockbackX"`    // knockback direction
	KnockbackY   float64          `json:"knockbackY"`
}

// WeaponDamageEvent is emitted when a weapon deals damage.
type WeaponDamageEvent struct {
	AttackerID   string     `json:"attackerId"`
	TargetID     string     `json:"targetId"`
	WeaponType   WeaponType `json:"weaponType"`
	Damage       float64    `json:"damage"`
	IsCritical   bool       `json:"isCritical"`
	IsDOT        bool       `json:"isDot"`
	IsLifesteal  bool       `json:"isLifesteal"`
	HealAmount   float64    `json:"healAmount,omitempty"`
	TargetX      float64    `json:"targetX"`
	TargetY      float64    `json:"targetY"`
}

File: packages/shared/src/types/weapons.ts
/**
 * v14 Phase 2: Megabonk Weapon System — Shared Types
 * Mirror of server/internal/domain/weapons.go
 */

// ─── Weapon Types ───

export type WeaponType =
  | 'bonk_mallet'
  | 'chain_bolt'
  | 'flame_ring'
  | 'frost_shards'
  | 'shadow_strike'
  | 'thunder_clap'
  | 'venom_cloud'
  | 'crystal_shield'
  | 'gravity_bomb'
  | 'soul_drain';

export const ALL_WEAPON_TYPES: WeaponType[] = [
  'bonk_mallet', 'chain_bolt', 'flame_ring', 'frost_shards',
  'shadow_strike', 'thunder_clap', 'venom_cloud', 'crystal_shield',
  'gravity_bomb', 'soul_drain',
];

export type WeaponPattern =
  | 'fan_swing'
  | 'chain_target'
  | 'circular_ring'
  | 'multi_shot'
  | 'teleport_back'
  | 'targeted_aoe'
  | 'deployable'
  | 'orbital'
  | 'deploy_explode'
  | 'beam';

export type SpecialEffect =
  | 'none'
  | 'knockback'
  | 'chain_decay'
  | 'burn_dot'
  | 'slow'
  | 'backstab'
  | 'stun'
  | 'poison_dot'
  | 'reflect'
  | 'pull'
  | 'lifesteal';

// ─── Weapon Data ───

export interface WeaponData {
  type: WeaponType;
  name: string;
  description: string;
  baseDPS: number;
  range: number;
  cooldownSec: number;
  pattern: WeaponPattern;
  specialEffect: SpecialEffect;
  // Pattern-specific parameters
  fanAngleDeg?: number;
  chainCount?: number;
  chainDecay?: number;
  projectileCount?: number;
  knockbackPx?: number;
  dotDurationSec?: number;
  dotDPSPerSec?: number;
  slowPercent?: number;
  slowDurationSec?: number;
  stunDurationSec?: number;
  deployDurationSec?: number;
  orbitalCount?: number;
  pullDurationSec?: number;
  lifestealPct?: number;
  backstabMult?: number;
}

// ─── Weapon Evolution ───

export interface WeaponEvolutionData {
  level: number;
  dpsMult: number;
  rangeMult: number;
  cooldownMult: number;
  ultimateName?: string;
  ultimateDesc?: string;
}

export interface WeaponSlot {
  type: WeaponType;
  level: number;        // 1~5
  cooldownTicks: number; // remaining cooldown
}

export const MAX_WEAPON_SLOTS = 5;
export const MAX_WEAPON_LEVEL = 5;

// ─── Static Weapon Definitions ───

export const ALL_WEAPONS: WeaponData[] = [
  {
    type: 'bonk_mallet', name: 'Bonk Mallet',
    description: 'Smashes enemies in a 120 degree frontal cone with knockback',
    baseDPS: 25, range: 80, cooldownSec: 1.5,
    pattern: 'fan_swing', specialEffect: 'knockback',
    fanAngleDeg: 120, knockbackPx: 30,
  },
  {
    type: 'chain_bolt', name: 'Chain Bolt',
    description: 'Lightning bolt chains through 3 enemies with decreasing damage',
    baseDPS: 15, range: 150, cooldownSec: 1.2,
    pattern: 'chain_target', specialEffect: 'chain_decay',
    chainCount: 3, chainDecay: 0.20,
  },
  {
    type: 'flame_ring', name: 'Flame Ring',
    description: 'Expanding 360 degree ring of fire that burns enemies',
    baseDPS: 20, range: 120, cooldownSec: 3.0,
    pattern: 'circular_ring', specialEffect: 'burn_dot',
    dotDurationSec: 2.0, dotDPSPerSec: 10,
  },
  {
    type: 'frost_shards', name: 'Frost Shards',
    description: 'Fires 5 ice shards in a 45 degree cone, slowing enemies',
    baseDPS: 12, range: 180, cooldownSec: 1.0,
    pattern: 'multi_shot', specialEffect: 'slow',
    fanAngleDeg: 45, projectileCount: 5,
    slowPercent: 0.40, slowDurationSec: 1.0,
  },
  {
    type: 'shadow_strike', name: 'Shadow Strike',
    description: 'Teleports behind the nearest enemy for a backstab',
    baseDPS: 40, range: 200, cooldownSec: 4.0,
    pattern: 'teleport_back', specialEffect: 'backstab',
    backstabMult: 2.0,
  },
  {
    type: 'thunder_clap', name: 'Thunder Clap',
    description: 'AOE lightning on the highest-HP enemy, stunning targets',
    baseDPS: 30, range: 250, cooldownSec: 3.5,
    pattern: 'targeted_aoe', specialEffect: 'stun',
    stunDurationSec: 0.5,
  },
  {
    type: 'venom_cloud', name: 'Venom Cloud',
    description: 'Deploys a poison cloud at current position',
    baseDPS: 8, range: 100, cooldownSec: 5.0,
    pattern: 'deployable', specialEffect: 'poison_dot',
    deployDurationSec: 5.0, dotDurationSec: 5.0, dotDPSPerSec: 8,
  },
  {
    type: 'crystal_shield', name: 'Crystal Shield',
    description: '3 orbiting crystals that damage on contact and reflect',
    baseDPS: 10, range: 50, cooldownSec: 0,
    pattern: 'orbital', specialEffect: 'reflect',
    orbitalCount: 3,
  },
  {
    type: 'gravity_bomb', name: 'Gravity Bomb',
    description: 'Deploys a black hole that pulls enemies in, then explodes',
    baseDPS: 35, range: 200, cooldownSec: 6.0,
    pattern: 'deploy_explode', specialEffect: 'pull',
    pullDurationSec: 2.0,
  },
  {
    type: 'soul_drain', name: 'Soul Drain',
    description: 'Continuous beam to nearest enemy, stealing HP',
    baseDPS: 18, range: 120, cooldownSec: 0,
    pattern: 'beam', specialEffect: 'lifesteal',
    lifestealPct: 0.30,
  },
];

// ─── Evolution Table ───

export const WEAPON_EVOLUTION_TABLE: WeaponEvolutionData[] = [
  { level: 1, dpsMult: 1.0, rangeMult: 1.0, cooldownMult: 1.0 },
  { level: 2, dpsMult: 1.3, rangeMult: 1.0, cooldownMult: 1.0 },
  { level: 3, dpsMult: 1.3, rangeMult: 1.25, cooldownMult: 1.0 },
  { level: 4, dpsMult: 1.3, rangeMult: 1.25, cooldownMult: 0.80 },
  { level: 5, dpsMult: 1.6, rangeMult: 1.5, cooldownMult: 0.70 },
];

// ─── Ultimate Names ───

export const ULTIMATE_WEAPON_NAMES: Record<WeaponType, string> = {
  bonk_mallet: 'Earthquake',
  chain_bolt: 'Storm Network',
  flame_ring: 'Inferno',
  frost_shards: 'Blizzard',
  shadow_strike: 'Phantom Dance',
  thunder_clap: 'Judgment',
  venom_cloud: 'Plague',
  crystal_shield: 'Diamond Fortress',
  gravity_bomb: 'Singularity',
  soul_drain: 'Life Siphon',
};

export const ULTIMATE_WEAPON_DESCS: Record<WeaponType, string> = {
  bonk_mallet: '360 degree shockwave + crack zone',
  chain_bolt: '5-chain + residual lightning field',
  flame_ring: 'Double ring + piercing',
  frost_shards: '360 degree ice storm',
  shadow_strike: 'Triple teleport + clone',
  thunder_clap: 'Map-wide lightning + 5s stun (once per epoch)',
  venom_cloud: 'Moving cloud + infection spread',
  crystal_shield: '5 orbitals + 200% reflect',
  gravity_bomb: '5s black hole + 2x radius',
  soul_drain: 'Frontal cone + ally heal',
};

// ─── Status Effects ───

export type StatusEffectType = 'burn' | 'poison' | 'slow' | 'stun' | 'pull' | 'knockback_effect';

export interface StatusEffect {
  type: StatusEffectType;
  sourceId: string;
  ticksLeft: number;
  damagePerTick: number;
  slowFraction: number;
  knockbackX: number;
  knockbackY: number;
}

// ─── Weapon Damage Event (for client rendering) ───

export interface WeaponDamageEvent {
  attackerId: string;
  targetId: string;
  weaponType: WeaponType;
  damage: number;
  isCritical: boolean;
  isDot: boolean;
  isLifesteal: boolean;
  healAmount?: number;
  targetX: number;
  targetY: number;
}

// ─── Weapon Colors for VFX ───

export const WEAPON_COLORS: Record<WeaponType, { primary: string; secondary: string }> = {
  bonk_mallet:    { primary: '#C0804060', secondary: '#FFD700' },
  chain_bolt:     { primary: '#00BFFF', secondary: '#FFFFFF' },
  flame_ring:     { primary: '#FF4500', secondary: '#FFD700' },
  frost_shards:   { primary: '#87CEEB', secondary: '#FFFFFF' },
  shadow_strike:  { primary: '#4B0082', secondary: '#8A2BE2' },
  thunder_clap:   { primary: '#FFD700', secondary: '#FFFFFF' },
  venom_cloud:    { primary: '#32CD32', secondary: '#006400' },
  crystal_shield: { primary: '#00CED1', secondary: '#E0FFFF' },
  gravity_bomb:   { primary: '#4B0082', secondary: '#000000' },
  soul_drain:     { primary: '#8B0000', secondary: '#FF6347' },
};

// ─── Helper Functions ───

export function getWeaponData(type: WeaponType): WeaponData | undefined {
  return ALL_WEAPONS.find(w => w.type === type);
}

export function getEvolutionData(level: number): WeaponEvolutionData | undefined {
  if (level < 1 || level > MAX_WEAPON_LEVEL) return undefined;
  return WEAPON_EVOLUTION_TABLE[level - 1];
}

export function getEffectiveDPS(weapon: WeaponData, level: number): number {
  const evo = getEvolutionData(level);
  if (!evo) return weapon.baseDPS;
  return weapon.baseDPS * evo.dpsMult;
}

export function getEffectiveRange(weapon: WeaponData, level: number): number {
  const evo = getEvolutionData(level);
  if (!evo) return weapon.range;
  return weapon.range * evo.rangeMult;
}

export function getEffectiveCooldown(weapon: WeaponData, level: number): number {
  const evo = getEvolutionData(level);
  if (!evo) return weapon.cooldownSec;
  return weapon.cooldownSec * evo.cooldownMult;
}
